README.md
# GLaDOS

![Who are you?](https://i.imgur.com/0gbb1AH.png)


## What is GLaDOS?

GLaDOS is an application for:
* automatic, continuous scanning of Valve TF2 servers to detect malicious bots
* providing an API for aggregated player/bot statistics and automatic [TF2BD](https://github.com/PazerOP/tf2_bot_detector) player/rules list generation
* directing hathook anti-bot map queueing to more efficiently target malicious bots
* receiving [hathook](https://github.com/incontestableness/hathook-public) bot events via [portalgun](https://github.com/incontestableness/portalgun)

GLaDOS also enables users running [hathook](https://github.com/incontestableness/hathook-public) to automatically abandon matches with malicious bots in them without having to join the server first. Similarly, hathook bots also take advantage of this functionality to avoid joining matches that don't have malicious bots to kick.

It is intended to be run on a centralized server as a WSGI application for optimal performance. GLaDOS is experimental software still in alpha.


## How do I use this?

**You don't need to run your own instance of GLaDOS.** A server is already hosted and configured as the default in [portalgun](https://github.com/incontestableness/portalgun).

Simply run a local instance of [portalgun](https://github.com/incontestableness/portalgun) and your [hathook](https://github.com/incontestableness/hathook-public) installation should work as intended.

**It is strongly recommended that you use the default central server as GLaDOS takes up considerable resources when scanning TF2 servers.**


## Is the GLaDOS API publicly available?

Yes! You can access the API at https://milenko.ml/api/something where "/something" is the relevant API call.

All responses are returned in JSON format, unless an error is encountered. Most calls are cached for about 5 seconds.

Here are the publicly available API calls:
* [/popmaps/\<region_id>](https://milenko.ml/api/popmaps/0) - A list of the most popular maps for malicious bots. See [regions.json](https://github.com/incontestableness/GLaDOS/blob/master/regions.json) for available region IDs.
* [/botnames](https://milenko.ml/api/botnames) - A list of **potential** bot names, based on multiple players using the same name. This endpoint provides every name that has been seen with an (N) prefix in the past 24 hours. Use the namerules endpoint if you want times_seen limits applied and dynamically created rules that cover variants of names.
* [/namerules](https://milenko.ml/api/namerules) - An automatically generated [TF2BD](https://github.com/PazerOP/tf2_bot_detector)-compatible rules list for current bot names based on scanning Valve's TF2 servers for patterns in names.
* [/check/\<server_address>](https://milenko.ml/api/check/208.78.165.231:27015) - The number of bots on the server and a list of bot names that might not have been otherwise detected (namestealers and char-injected names).
* [/stats](https://milenko.ml/api/stats) - Cumulative statistics.


## Why did you do this?

1. In order to enable hosting anti-bots in multiple regions while still gathering aggregated statistics easily
2. Much more efficient targeting of matches with malicious bots
3. Provide for a better [hathook](https://github.com/incontestableness/hathook-public) user experience
4. Create an API that enables displaying statistics on [the website](https://milenko.ml/)
5. Science is fun


## Why the name?

["It's funny, actually, when you think about it."](https://i1.theportalwiki.net/img/3/3b/GLaDOS_escape_01_part1_nag09-1.wav) - Potato GLaDOS


CHANGELOG.md
# GLaDOS Changelog

## [3.1.0] - 2021-07-04
### First "stable" alpha release
+ Dynamic configuration options
+ Soft restart support to preserve API data when restarting GLaDOS for development and maintenance
+ Removed API ratelimits
+ Reduced API cache times, ultimately down to 5 seconds
+ Added `getnames.py` for viewing potential bot names from the API and detecting evasion characters in bot names

+ Implemented detection of namestealing bots
+ Improved server scanning performance, ultimately making scans take less than 5 seconds
+ Implemented fully automatic bot name detection with automatic blacklisting and improved bot name matching

+ Added HTTPS support to the website and the GLaDOS API

+ Created a multithreaded and asynchronous program (`ketchupbottle.py`) based on `zeolite.py` to find Valve servers not reported by the `ISteamApps/GetServersAtAddress` Steam API
+ Rewrote GLaDOS to use the much more reliable `IGameServersService/GetServerList` Steam API, deprecating the three previous methods of host discovery (thanks to Makamoto from teamwork.tf for the info!)

+ Increased reliability of data persistence by autosaving periodically
+ Restructured the GLaDOS source code to be more readable and maintainable
+ Enhanced debug logging, including better control via the `settings.ini` file
+ Repeated improvements to several of the GLaDOS API calls, including new features and better human readability/useability

+ Created the `valve_whitelist.py` module to dynamically generate a set of CIDR ranges comprising datacenters in Valve's `NetworkDatagramConfig.json` and Valve's ASN
+ Implemented Valve IP/port whitelist validation for the `/check/<server>` API

+ Finally implemented bot name variant tracking and multi-pattern matching in the `/namerules` API
  + This allows GLaDOS to manage bot names with invisible unicode characters and the like as well as for the rules to detect them reliably
+ Finally implemented asynchronous server scanning, making scans even faster than before
  + Added dynamic times seen requirements based on scan frequency, to ensure namerules accuracy
  + Added a scan frequency stabilizer to prevent GLaDOS from scanning too often
+ Fixed issues with namerules regex not being created correctly
+ Restructured the codebase (again) to tidy things up
+ Way too many bug fixes and miscellaneous improvements to list

## [3.0.0] - 2021-04-19
### Initial alpha release
+ Basic support for running as a WSGI application
+ Basic support for scanning servers in lists generated by `potato.py`
+ Detection of malicious bots based on manually defined regex patterns in `name_blacklist.txt`
+ `/popmaps/<desired_region>` API for seeing which maps malicious bots have recently been seen on
+ `/botnames` API for seeing potential malicious bot names
+ `/check/<server>` API for retrieving the number of detected malicious bots on any given server


GLaDOS.py
#!/usr/bin/env python3

__version__ = "3.1.0"

import os

# For running as a WSGI application, as it should be in production
if __name__ != "__main__":
	os.chdir(os.path.expanduser("~/GLaDOS"))

import argparse
from debug import lc
from flask import abort, Flask, jsonify, redirect, request
from flask_caching import Cache
from helpers import matchify
import json
import logging
from logging.handlers import RotatingFileHandler
from moralitycore import MoralityCore
import netaddr
import re
import socket
import time
from valve_whitelist import cidrs



# Configuration options
parser = argparse.ArgumentParser()
# You may wish to tune this based on your ping to the servers
parser.add_argument("--scan-timeout", type=float, default=0.500)
# How frequently a scan runs, in seconds. Affects times seen requirements and delay between scans.
parser.add_argument("--scan-frequency", type=float, default=5.0)
args = parser.parse_args()


# Logging setup
logging.basicConfig(
        handlers = [RotatingFileHandler("log.txt", maxBytes=1024 * 1024 * 20, backupCount=2)],
        level = logging.INFO,
        format = "[%(asctime)s] [%(funcName)s:%(lineno)d] %(message)s",
        datefmt = "%a %b %d @ %R:%S")
logger = logging.getLogger()
print = logger.info


# Calculate appropriate times seen requirements based on scanning speed
# Reasonably confident. For a false positive, a single user with a name starting with (N) would have to be in-game for the full course of an hour.
suspicious_times_seen = int((60 * 60) / args.scan_frequency)
# Very high confidence. For a false positive, a single user with a name starting with (N) would have to be in-game for the full course of 3 hours.
cheater_times_seen = int((60 * 60 * 3) / args.scan_frequency)
print(f"Times seen requirements: {suspicious_times_seen} for suspicious, {cheater_times_seen} for cheater")


# Read in regions.json for nicer stats output
file = open("regions.json", "r")
data = file.read()
file.close()
regions_data = json.loads(data)


# Create a GLaDOS core
core = MoralityCore(args.scan_frequency, args.scan_timeout, suspicious_times_seen, cheater_times_seen)


# Declare our Flask instance and define all the API routes
api = Flask(__name__)
config = {
	"JSON_SORT_KEYS": False,
	"CACHE_TYPE": "SimpleCache", # Flask-Caching
	"CACHE_DEFAULT_TIMEOUT": 5
}
api.config.from_mapping(config)
cache = Cache(api)


def whitelisted():
	with open("ip_whitelist.txt", "r") as wl:
		whitelist = wl.read().split("\n")[:-1]
		return request.remote_addr in whitelist


# Redirect to static content about the API
@api.route("/")
@cache.cached(timeout=60 * 60)
def root():
	return redirect(location="/api.html")


# We can load an updated regex name blacklist on demand without restarting
# The same goes for logging settings
# The requesting IP must be whitelisted in ip_whitelist.txt
@api.route("/reload")
def reload():
	if not whitelisted():
		abort(403)
	lc.reload_settings()
	core.load_blacklists()
	return jsonify({"response": {"success": True}})


# Save current data and terminate the backend.
# Useful for using new code without re-collecting data.
@api.route("/restart")
def restart():
	if not whitelisted():
		abort(403)
	# Ask the core to stop doing stuff
	core.halt = True
	# Wait
	while not core.halted:
		time.sleep(0.100)
	# Save stuff
	core.save_data()
	# Crash
	os.abort()


# Return a list of map names to target
@api.route("/popmaps/<desired_region>")
@cache.cached(forced_update=whitelisted)
def popmaps(desired_region):
	targeted = []
	tracker = core.region_map_trackers[int(desired_region)]
	for i in tracker["popular_bot_maps"]:
		targeted.append(i.name)
	return jsonify({"response": {"popular_bot_maps": targeted}})


# Return a list of bot names we have seen while scanning
# The counter accumulates, which makes it useful for measuring "presence" over time
# A bot will have a higher score at the end of the measurement period
@api.route("/botnames")
@cache.cached(forced_update=whitelisted)
def botnames():
	bot_names = []
	now = time.time()
	for pn in sorted(core.bot_names.values(), reverse=True):
		# Ignore names older than 24h
		if now - pn.last_seen >= 60 * 60 * 24:
			continue
		json_quote_escaped = pn.name.replace("\"", "\\\"")
		bot_names.append({
			"name": json_quote_escaped,
			"properties": {
				"times_seen": pn.times_seen,
				"first_seen": f"{int((now - pn.first_seen) / 60)} minutes ago",
				"last_seen": f"{int((now - pn.last_seen) / 60)} minutes ago"
			}
		})
	return jsonify({"response": {"bot_names": bot_names}})


# "I know you're there. I can feel you here."
# Hathook users and bots can automatically accept/reject assigned matches depending on whether they have bots, before joining them
# portalgun will convert the namestealers list into a format hathook can process safely
# Namestealing bots will be marked as soon as users/bots join the match
# A timeout of 3 seconds would give more recent data upon multiple calls but 30 would ensure stability. Play it safe for now.
@api.route("/check/<server>")
@cache.cached(timeout=30)
def user_checkServer(server):
	if ":" not in server:
		server = f"{server}:27015"
	ip, port = server.split(":")
	# Is this a Valve IP?
	ip = netaddr.IPAddress(ip)
	if not any(ip in cidr for cidr in cidrs):
		return jsonify({"response": {"success": False, "message": "IP address does not appear to be a Valve IP"}})
	# Is this a reasonable port?
	if int(port) not in range(27015, 27270 + 1):
		return jsonify({"response": {"success": False, "message": "Port not allowed"}})
	try:
		bot_count, namestealers = core.checkServer(server)
	except socket.timeout:
		return jsonify({"response": {"success": False, "message": "Connection timed out"}})
	return jsonify({"response": {"success": True, "server": server, "bot_count": bot_count, "namestealers": namestealers}})


# TODO: Statistics events
@api.route("/event/exampleEvent/<data>")
def exampleEventHandler(data):
	if not whitelisted:
		abort(403)
	return jsonify({})


# TODO: Cumulative statistics from bot events
@api.route("/stats")
@cache.cached(forced_update=whitelisted)
def stats():
	casual_total = 0
	bot_total = 0
	players_per_region = {}
	bots_per_region = {}
	for region_id in core.region_map_trackers:
		tracker = core.region_map_trackers[region_id]
		region_simple_name = regions_data[f"Region ID {region_id}"]["simple_name"]
		casual_total += tracker["casual_in_game"]
		bot_total += tracker["malicious_in_game"]
		players_per_region[region_simple_name] = tracker["casual_in_game"]
		bots_per_region[region_simple_name] = tracker["malicious_in_game"]
	return jsonify({"response": {
			"casual_in_game": {
				"totals": {"all_players": casual_total, "malicious_bots": bot_total},
				"per_region": {"all_players": players_per_region, "malicious_bots": bots_per_region}
			}
		}
	})


@api.route("/debug")
def debug():
	patterns = [pat.pattern for pat in core.patterns]
	return jsonify({"response": {"debug_info": {"version": __version__, "patterns": patterns}}})


# "You picked the wrong house, fool!"
# Automatic TF2BD rules list creation based on current common bot names
@api.route("/namerules")
@cache.cached(forced_update=whitelisted)
def namerules():
	bnames = []
	snames = []
	for pn in sorted(core.bot_names.values(), reverse=True):
		# Ignore names older than 24h
		if time.time() - pn.last_seen >= 60 * 60 * 24:
			continue
		if pn.times_seen >= cheater_times_seen:
			bnames.append(pn)
		elif pn.times_seen >= suspicious_times_seen:
			snames.append(pn)
	data = {"$schema": "https://raw.githubusercontent.com/PazerOP/tf2_bot_detector/master/schemas/v3/rules.schema.json"}
	data["file_info"] = {
		"authors": ["Sydney", "The Great Milenko"],
		"title": "Vibe check",
		"description": "GLaDOS automatic bot name detection rules list",
		"update_url": "https://milenko.ml/api/namerules"
	}
	rules = []
	dup_prefix_regex = "(\\([1-9]\\d?\\))?"
	for pn in bnames:
		rule = {"actions": {"transient_mark": ["cheater"]}}
		patterns = [f"{dup_prefix_regex}{matchify(pn.name)}"]
		for variant in pn.variants:
			patterns.append(f"{dup_prefix_regex}{matchify(variant)}")
		rule["description"] = f"\"{pn.name}\" seen {pn.times_seen} times in 24h"
		rule["triggers"] = {
			"username_text_match": {
				"case_sensitive": True,
				"mode": "regex",
				"patterns": patterns
			}
		}
		rules.append(rule)
	for pn in snames:
		rule = {"actions": {"transient_mark": ["suspicious"]}}
		patterns = [f"{dup_prefix_regex}{matchify(pn.name)}"]
		for variant in pn.variants:
			patterns.append(f"{dup_prefix_regex}{matchify(variant)}")
		rule["description"] = f"\"{pn.name}\" seen {pn.times_seen} times in 24h"
		rule["triggers"] = {
			"username_text_match": {
				"case_sensitive": True,
				"mode": "regex",
				"patterns": patterns
			}
		}
		rules.append(rule)
	data["rules"] = rules
	response = jsonify(data)
	response.headers["Content-Disposition"] = "inline; filename=rules.GLaDOS.json"
	return response


# Run the GLaDOS API server in the main thread, if running directly for development/testing and not as a WSGI application
if __name__ == "__main__":
	api.run(port=8000)
